File: sol/solution/0101-0200/0105/Solution.dart
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */
class Solution {
  TreeNode? buildTree(List<int> preorder, List<int> inorder) {
    if (preorder.isEmpty || inorder.isEmpty) return null;
    Map<int, int> inorderIndex = {};
    for (int i = 0; i < inorder.length; i++) {
      inorderIndex[inorder[i]] = i;
    }
    TreeNode? helper(int preStart, int inStart, int inEnd) {
      if (preStart >= preorder.length || inStart > inEnd) return null;
      int rootVal = preorder[preStart];
      TreeNode root = TreeNode(rootVal);
      int idx = inorderIndex[rootVal]!;
      int leftSize = idx - inStart;
      root.left = helper(preStart + 1, inStart, idx - 1);
      root.right = helper(preStart + leftSize + 1, idx + 1, inEnd);
      return root;
    }
    return helper(0, 0, inorder.length - 1);
  }
}

File: sol/solution/0101-0200/0105/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */
object Solution {
  def buildTree(preorder: Array[Int], inorder: Array[Int]): TreeNode = {
    if (preorder.isEmpty || inorder.isEmpty) return null
    val inMap = inorder.zipWithIndex.toMap
    def helper(preStart: Int, preEnd: Int, inStart: Int, inEnd: Int): TreeNode = {
      if (preStart > preEnd || inStart > inEnd) return null
      val rootVal = preorder(preStart)
      val root = new TreeNode(rootVal)
      val inRootIdx = inMap(rootVal)
      val leftSize = inRootIdx - inStart
      root.left = helper(preStart + 1, preStart + leftSize, inStart, inRootIdx - 1)
      root.right = helper(preStart + leftSize + 1, preEnd, inRootIdx + 1, inEnd)
      root
    }
    helper(0, preorder.length - 1, 0, inorder.length - 1)
  }
}